### PR TITLE
HIP: fix ArithTraits to support HIP backend, see issue #807

### PR DESCRIPTION
**Pull request by lucbv**
The goal is to replace undefined calls to math functions with the HIP backend.
Hopefully the pre-processor logic makes sense?

### CODE REVIEW

File: src/Kokkos_ArithTraits.hpp [hunk 1]
```diff
@@ -1126,9 +1128,11 @@ class ArithTraits<double> {
     return ::atan (x);
   }
   static KOKKOS_FORCEINLINE_FUNCTION val_type nan () {
-#ifdef __CUDA_ARCH__
+#if defined(__CUDA_ARCH__)
     return CUDART_NAN;
     //return nan (); // this returns 0 ???
+#elif defined(__HIP_DEVICE_COMPILE__)
+    return nan("");
```
Comment (dalg24) [inline src/Kokkos_ArithTraits.hpp:1135]: I get
```
<kokkos-kernels>/src/Kokkos_ArithTraits.hpp:1135:12: error: too many arguments to function call, expected 0, have 1; did you mean '::nan'?
    return nan("");
           ^~~
           ::nan
/opt/rocm-3.5.0/llvm/lib/clang/11.0.0/include/__clang_cuda_math_forward_declares.h:141:19: note: '::nan' declared here
__DEVICE__ double nan(const char *);
```

Comment (dalg24) [reply src/Kokkos_ArithTraits.hpp:1135]:
```suggestion
    return std::nan("");
```

**Comment by lucbv** [comment]
OK got the PR is building correctly on `caraway` and I checked that the spot-check runs fine on kokkos-dev2.
I think this is good for a final check before merging, @srajama1 any thoughts?

**Review by srajama1 — APPROVED** [review]
The changes look benign to me. However, I would like feedback from someone who has experience with this software stack. @dalg24 already commented on this PR, if he approves go for it.

File: src/Kokkos_ArithTraits.hpp [hunk 2]
```diff
@@ -1132,7 +1132,7 @@ class ArithTraits<double> {
     return CUDART_NAN;
     //return nan (); // this returns 0 ???
 #elif defined(__HIP_DEVICE_COMPILE__)
-    return nan("");
+    return ::nan("");
```
Comment (dalg24) [inline src/Kokkos_ArithTraits.hpp:1135]: Nope you need `std::`

**Review by dalg24 — APPROVED** [review]

**Comment by srajama1** [comment]
Merging it. Thanks @lucbv !

**Pull request merged**